src/lib/ecore/efl_promise2.c

```diff
@@ -0,0 +1,277 @@
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "Ecore.h"
+#include "efl_promise2.h"
+#include "ecore_private.h"
+#include "efl_promise2_priv.h"
+
+struct _Efl_Promise2 {
+   Eina_Free_Cb cancel;
+   Efl_Future2 *future;
+   const void *data;
+};
+
+struct _Efl_Future2 {
+   Efl_Future2_Cb success;
+   Efl_Future2_Cb error;
+   Eina_Free_Cb free;
+   Efl_Promise2 *promise;
+   Efl_Future2 *next;
+   Efl_Future2 *prev;
+   Eina_Value *pending_value;
+   const void *data;
+   Ecore_Event *event;
+   Eina_Bool reject;
+};
+
+static Eina_Mempool *_promise_mp = NULL;
+static Eina_Mempool *_future_mp = NULL;
+static Eina_List *_pending_futures = NULL;
+static int FUTURE_EVENT_ID = -1;
+static Ecore_Event_Handler *_future_event_handler = NULL;
+
+static void
+_efl_future2_dispatch(Efl_Future2 *f, Eina_Value *value, Eina_Bool rejected)
+{
+   while (f)
+     {
+        if (((!rejected && f->success) || (rejected && f->error)))
+          break;
+        Efl_Future2 *n = f->next;
+        Eina_Free_Cb free_cb = f->free;
+        const void *data = f->data;
+        if (n) n->prev = NULL;
+        f->next = NULL;
+        eina_mempool_free(_future_mp, f);
+        if (free_cb) free((void *)data);
+        f = n;
+     }
+
+   if (!f)
+     {
+        eina_value_free(value);
+        return;
+     }
+
+   Eina_Value *next_value = rejected ? f->error((void *)f->data, value) :
+     f->success((void *)f->data, value);
+   if (value != next_value) eina_value_free(value);
+   if (f->free) f->free((void *)f->data);
```

Comment:
here the callbacks may lead user to change `f`, for instance `efl_future_cancel(f)` from one of the callbacks, not smart, but we shouldn't crash... then either store these pointers in the stack, or flag as "dispatching" so cancel will just ignore the request.

since we want to be tight on memory, don't add a boolean as it will grow the struct size by 7 bytes on 64bits... instead just change `success/error` prior to calling them:

```c
Efl_Future2_Cb cb = rejected ? f->error : f->success;
f->error = f->success = EFL_FUTURE2_DISPATCHED; // ie: (Efl_Future2_Cb)0x01
Eina_Value *next_value = cb((void *)f->data, value);
...
```
and inside `efl_future_cancel()` you check `f->success == EFL_FUTURE2_DISPATCHED`...